src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb

```diff
@@ -81,6 +81,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
         Private ReadOnly _syntaxTrees As ImmutableArray(Of SyntaxTree)
 
         ''' <summary>
+        ''' The syntax trees explicitly given to the compilation at creation, in set form.
+        ''' </summary>
+        Private ReadOnly _syntaxTreeSet As HashSet(Of SyntaxTree)
```

Comment:
Also, take a look at MyTemplateLocation and EmbeddedTreeLocation, these are the locations created for the non-user trees. If we have a location, it might be even faster to get an answer from it. Perhaps IsInSource property returns false.